mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/options.scala

```diff
@@ -234,3 +246,295 @@ object SMTEncoding {
     case oddEncodingType => throw new IllegalArgumentException(s"Unexpected SMT encoding type $oddEncodingType")
   }
 }
+
+/** The basic interface for groups of options used to control program behavior */
+sealed trait OptionGroup
+
+/**
+ * Specifies the options used to control pass executions
+ *
+ * Typically, each pass will need a subset of the available option groups, and the sequence of passes run by the
+ * `PassExecutor` will require the union of the option groups required by its constituent passes.
+ *
+ * The unions can be specified via the `Has*` traits, such as [[OptionGroup.HasInput]] and constructed via the `With*`
+ * classes, such as [[OptionGroup.WithInput]].
+ */
+object OptionGroup {
+
+  /**
+   * Interface for a group of related options that can be be produced from a [[Config]]
+   *
+   * The intended use of this class is to identify '''configurable''' options. Configurable options are extensions of
+   * `OptionGroup` that can be derived from a `Config`. Typically, these are case classes.
+   *
+   * @param cfg
+   *   An instance of [[Config config group]].
+   */
+  sealed trait Configurable[C <: Config[C], O] {
+    //  TODO could manual apply methods be replaced with pureconfig merging?
+    def apply(cfg: C): Try[O]
+  }
+
+  // Convert optional values into `Try`'s
+  // see https://stackoverflow.com/questions/17521709/how-can-i-best-convert-an-option-into-a-try/45017589#45017589
+  implicit class OptionOps[A](opt: Option[A]) {
+
+    def toTry(field: String): Try[A] = {
+      opt
+        .map(Success(_))
+        .getOrElse(Failure(new PassOptionException(s"Missing value for required option ${field}")))
+    }
+  }
+
+  /** Options used in all modes of execution */
+  case class Common(
+      command: String,
+      debug: Boolean,
+      features: Seq[Feature],
+      outDir: File,
+      profiling: Boolean,
+      runDir: Option[File],
+      smtprof: Boolean,
+      writeIntermediate: Boolean)
+      extends OptionGroup
```

Comment:
So at this point adding new options becomes a bit expensive with boilerplate in multiple places. I'm wondering if we cannot recycle the `Config` objects and dynamically ensure that the required options are set?

Reply:
Right. I wanted to be sure people were ok with this before doing most of the work, which is why I asked in our stand ups :)

In the most common case, 3 entries are required:

- The cli arg
- The the mapping to the config class
- The mapping to the option class

That should be about 5 changes all together, including adding the fields and maps. It is a fair bit of boilerplate.

keep in my that with the current system, 3 entries are also required to add a cli arg that can be set by configuration:

- cli arg
- mapping to config
- mapping to option map

Plus validation logic is needed at any access of non-optional values and we have to cast them to the right types at access time. So while I’m not thrilled about the amount of boiler plate, afaict this doesn’t seem to make matters worse? I *think* it’s about the same amount of overhead, only it makes it possible to reason about and get static guarantees. 

It does make the amount of overhead more apparent tho, because it puts it all in one place :)

I’ve considered reusing the config and have these questions around it:

- then do we give up on static guarantees around which subset of the configs are used? Or are you thinking we’d use the same interface pattern, still?
- How do we specify the required options and provide safe access to them without writing approximately the same amount of boilerplate? Or are you suggesting we just make each access of required options have responsibility for possibly crashing/throwing?
- We would lose the ability to introduce an abstraction barrier between options provided and config values supplied. We’re only taking advantage of this in a few minutes laces currently (with input config group and when m dropping the config file in the options), but to me it makes sense that the shape of the configurations users can supply will not in general be isomorphic to the shape of the options we need to run process. What are your thoughts on this trade off?

Reply:
I guess my concern stems from (a) the assumption that all the entries would be in the same place / file, and (b) a lack of understanding of how the configs will eventually be wired through. Maybe it's already obvious how to introduce these entries if one actually wants to access an option in code?

Reply:
> I _think_ it’s about the same amount of overhead

Not sure it's the same amount overhead; onboarding for a mutable map is easier than having to understand a custom class structure (not saying that I prefer the mutable map, though ;) )

> * then do we give up on static guarantees around which subset of the configs are used? Or are you thinking we’d use the same interface pattern, still?

Anything preventing us from using it?

> * How do we specify the required options and provide safe access to them without writing approximately the same amount of boilerplate? Or are you suggesting we just make each access of required options have responsibility for possibly crashing/throwing?

My understanding was that even non-required options have some kind of default value. I would assume that only fully populated `Config` objects reach the actual option-consuming code.

> * We would lose the ability to introduce an abstraction barrier between options provided and config values supplied. We’re only taking advantage of this in a few minutes laces currently (with input config group and when m dropping the config file in the options), but to me it makes sense that the shape of the configurations users can supply will not in general be isomorphic to the shape of the options we need to run process. What are your thoughts on this trade off?

That's a good point. Currently we're not making much use of it? Also sounds to me like such "composite" options should be computed values, and the logic for them could reside inside the configuration classes?

Reply:
> Not sure it's the same amount overhead

I meant the overhead in terms of lines of code to write. The *think* the cognitive overhead here more or less follows the general tradeoff with static vs. dynamic typing: this approach moves some complexity into the types, but removes some complexity from the logic around values (e.g., you don't need validation and type casting logic at time of access).